src/include/migraphx/op/loop.hpp

```diff
@@ -0,0 +1,164 @@
+
+#include "migraphx/errors.hpp"
+#include "migraphx/raw_data.hpp"
+#include <array>
+#include <iterator>
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/module.hpp>
+#include <cmath>
+#include <string>
+#include <utility>
+#include <set>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct loop
+{
+    int64_t max_iter_num = 0;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.max_iter_num, "max_iter_num"));
+    }
+
+    std::string name() const { return "loop"; }
+
+    shape compute_shape(const std::vector<shape>& inputs, std::vector<module_ref> mods) const
+    {
+        check_shapes{inputs, *this}.standard();
+        if(mods.size() != 1)
+        {
+            MIGRAPHX_THROW("LOOP: operator should have one submodule.");
+        }
+
+        const auto& mod     = mods.front();
+        auto mod_out_shapes = mod->get_output_shapes();
+        auto dep_param_num  = inputs.size() - 2;
+        // first two names -- iter_num and cond_var -- are not counted
+        mod_out_shapes.erase(mod_out_shapes.begin());
+        std::vector<shape> ins_out_shapes(mod_out_shapes.begin(),
+                                          mod_out_shapes.begin() + dep_param_num);
+        mod_out_shapes.erase(mod_out_shapes.begin(), mod_out_shapes.begin() + dep_param_num);
+        for(const auto& out_s : mod_out_shapes)
+        {
+            auto lens = out_s.lens();
+            lens.insert(lens.begin(), max_iter_num);
+            ins_out_shapes.push_back({out_s.type(), lens});
+        }
+
+        return shape(ins_out_shapes);
+    }
+
+    std::pair<int, bool> get_name_index(const std::string& name,
+                                        const std::string& param_prefix) const
+    {
+        auto loc = name.find(param_prefix);
+        if(loc != std::string::npos)
+        {
+            int index = std::stoi(name.substr(loc + param_prefix.size()));
+            return {index, true};
+        }
+
+        std::string out_prefix = "#output_";
+        loc                    = name.find(out_prefix);
+        if(loc != std::string::npos)
+        {
+            int index = std::stoi(name.substr(loc + out_prefix.size()));
+            return {index, false};
+        }
+
+        return {-1, false};
+    }
+
+    argument compute(const shape& out_shape,
+                     const std::vector<argument>& args,
+                     const std::vector<module_ref>& mods,
+                     const std::function<std::vector<argument>(
+                         module_ref&, const std::unordered_map<std::string, argument>&)>& run) const
+    {
+        auto iter_num = args.at(0).at<int64_t>();
+        if(iter_num > max_iter_num)
+        {
+            MIGRAPHX_THROW("LOOP compute(): iter_num " + std::to_string(iter_num) +
```

Comment:
I think the only way this would happen is if the user manually creates a loop operator and sets the input arg for `iter_num` to be larger than `max_iter_num`. In onnx parsing, we are currently always setting `max_iter_num` to whatever is specified by either the input arg or the onnx options. Should we just make this an assert?